nupic/research/frameworks/continuous_learning/cl_utils.py

```diff
@@ -0,0 +1,256 @@
+    :param post_batch_callback: Callback function to be called after every batch
+                                with the following parameters: model, batch_idx
+    :type post_batch_callback: function
+    :param pre_batch_callback: Callback function to be called before every batch
+                               with the following parameters: model, batch_idx
+    :type pre_batch_callback: function
+    :param progress_bar: Optional :class:`tqdm` progress bar args.
+                         None for no progress bar
+    :type progress_bar: dict or None
+
+    :return: mean loss for epoch
+    :rtype: float
+    """
+    model.train()
+    # Use asynchronous GPU copies when the memory is pinned
+    # See https://pytorch.org/docs/master/notes/cuda.html
+    async_gpu = loader.pin_memory
+    if progress_bar is not None:
+        loader = tqdm(loader, **progress_bar)
+        # update progress bar total based on batches_in_epoch
+        if batches_in_epoch < len(loader):
+            loader.total = batches_in_epoch
+
+    # Check if training with Apex Mixed Precision
+    # FIXME: There should be another way to check if 'amp' is enabled
+    use_amp = hasattr(optimizer, "_amp_stash")
+    try:
+        from apex import amp
+    except ImportError:
+        if use_amp:
+            raise ImportError(
+                "Mixed precision requires NVIDA APEX."
+                "Please install apex from https://www.github.com/nvidia/apex")
+
+    t0 = time.time()
+    for batch_idx, (data, target) in enumerate(loader):
+        if batch_idx >= batches_in_epoch:
+            break
+
+        num_images = len(target)
+        data = data.to(device, non_blocking=async_gpu)
+        target -= 1
+        target = target.to(device, non_blocking=async_gpu)
+        t1 = time.time()
+
+        if pre_batch_callback is not None:
+            pre_batch_callback(model=model, batch_idx=batch_idx)
+
+        optimizer.zero_grad()
+        output = model(data)
+        loss = criterion(output, target)
+        del data, target, output
+
+        t2 = time.time()
+        if use_amp:
+            with amp.scale_loss(loss, optimizer) as scaled_loss:
+                scaled_loss.backward()
+        else:
+            loss.backward()
+
+        t3 = time.time()
+        optimizer.step()
+        t4 = time.time()
+
+        if post_batch_callback is not None:
+            time_string = ("Data: {:.3f}s, forward: {:.3f}s, backward: {:.3f}s,"
+                           + "weight update: {:.3f}s").format(t1 - t0, t2 - t1, t3 - t2,
+                                                              t4 - t3)
+            post_batch_callback(model=model, loss=loss.detach(), batch_idx=batch_idx,
+                                num_images=num_images, time_string=time_string)
+        del loss
+        t0 = time.time()
+
+    if progress_bar is not None:
+        loader.n = loader.total
+        loader.close()
+
+
+def unravel_index(index, shape):
+    """stole from
```

Comment:
`s/stole from/copied from/`

BTW, I cant find this function used anywhere, why do you need it?

Reply:
Oh right, it was for something else but torch.reshape() was more efficient. I'll get rid of it